File: python/leetcode8.py
class AtoiFSM:
    def __init__(self):
        self.state = "start"
        self.sign = 1
        self.val = 0
        self.table = {
                      #   space    +/-      0-9     other
            "start" :   ["start", "sign", "number", "end"],
            "sign"  :   ["end",   "end",  "number", "end"],
            "number":   ["end",   "end",  "number", "end"],
            "end"   :   ["end",   "end",  "end",    "end"], 
        }
    
    def _get_state_transfer_id(self, c):
        if c == " ":
            return 0
        elif c == "+" or c == "-":
            return 1
        elif c.isdigit():
            return 2
        else:
            return 3

    def run(self, c):
        # 状态转换
        self.state = self.table[self.state][self._get_state_transfer_id(c)]

        if self.state == "sign":
            self.sign = 1 if c == "+" else -1
        elif self.state == "number":
            self.val = self.val * 10 + int(c)
            self.val = min(self.val, 2**31-1) if self.sign == 1 else min(self.val, 2**31)
        elif self.state == "end":
            return self.sign * self.val
        return None


class Solution:
    def myAtoi(self, s):
        fsm = AtoiFSM()
        for c in s:
            res = fsm.run(c)
            if res is not None:
                return res
        return fsm.sign * fsm.val

File: java/leetcode8.java
import java.util.HashMap;
import java.util.Map;

class AtoiFSM {
    private String state = "start";
    public int sign = 1;
    public long val = 0;
    private Map<String, String[]> table = new HashMap<>();

    public AtoiFSM() {
        table.put("start", new String[]{"start", "sign", "number", "end"});
        table.put("sign", new String[]{"end", "end", "number", "end"});
        table.put("number", new String[]{"end", "end", "number", "end"});
        table.put("end", new String[]{"end", "end", "end", "end"});
    }

    private int getStateTransferID(char c){
        if (c == ' ') {
            return 0;
        }
        else if (c == '+' || c == '-') {
            return 1;
        }
        else if (Character.isDigit(c)){
            return 2;
        }
        return 3;
    }

    public void run(char c){
        state = table.get(state)[getStateTransferID(c)];
        if (state.equals("sign")) {
            sign = c == '+' ? 1 : -1;
        }
        else if (state.equals("number")){
            val = val * 10 + c - '0';
            val = sign == 1 ? Math.min(val, Integer.MAX_VALUE) : Math.min(val, - (long) Integer.MIN_VALUE);
        }
    }
}


class Solution {
    private AtoiFSM fsm = new AtoiFSM();
    public int myAtoi(String s) {
        for(int i = 0; i < s.length(); i++){
            fsm.run(s.charAt(i));
        }
        return (int) (fsm.sign * fsm.val);
    }
}